### ISSUE TITLE
surface to volume transformation?

### ISSUE THREAD

**Issue by PeerHerholz** [miykael/parcellation_fragmenter#32]
Hey folks,

I was just wondering, if a small function to transform the parcels from surface to volume
would be nice and useful, as a lot of people mainly work with volumes?
Maybe something like `annot2label` -> `mri_label2vol` should already work, eh?
Would be great to hear your thoughts on this.

Best, Peer

**Comment by miykael** [comment]
I think this is a very good idea! Like this, somebody can create the desired parcels but move it quickly back into volume space! I'm all for it :+1:!

**Comment by kristianeschenburg** [comment]
I think this is a good idea and useful -- @PeerHerholz do you want to create a pull request or should one of us?

**Comment by ctoroserey** [comment]
Sounds good and useful!

**Comment by PeerHerholz** [comment]
Coolio. 
I could work on it, but based on my current schedule it will take a while
till I get to it. If that's okay for you folks...

**Comment by PeerHerholz** [comment]
Ahoi hoi folks,

long long time ago...thx to @ctoroserey I'm on it again. 
In order to avoid having FreeSurfer as a dependency and remain in pure python land, we could use nilearn and nibabel to get the job done. How's everyone feeling about this?

**Comment by ctoroserey** [comment]
Yeah, I like that! I'd like to see how some of the parcel transformations look like before implementing it, though (I've had mixed results in the past).

**Comment by PeerHerholz** [comment]
Fair point, we would need to provide options for smoothing, etc. within the respective function. I'll try a few things and post updates here.

**Comment by PeerHerholz** [comment]
Ok, turns out there might be no pure python solution. Hence, I'll try [reprozip](https://www.reprozip.org/) again to get only the FreeSurfer components that are necessary.

**Comment by kristianeschenburg** [comment]
Dipy might have a solution to this... otherwise we'll need to make use of / implement some sort of interpolation ourselves

**Comment by PeerHerholz** [comment]
I briefly checked and that does not seem like it. I talked to @kaczmarj and he provided a nice solution to only include certain `FreeSurfer` functions. I'll give it a try with `annot2label` and `mri_surf2vol` and let you know.

**Comment by kaczmarj** [comment]
hi everyone, `mri_surf2vol` is a compiled executable, and most of its linked libraries are fairly standard. libgomp1 will probably need to be installed. here's the `ldd` output on my computer. i don't have `annot2label` on my computer for some reason.

```
jakub@dash:~$ ldd /usr/local/freesurfer/bin/mri_surf2vol
	linux-vdso.so.1 (0x00007ffd7dfb4000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f9272e34000)
	libcrypt.so.1 => /lib/x86_64-linux-gnu/libcrypt.so.1 (0x00007f9272dfa000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f9272df5000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f9272dd4000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f9272c50000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f9272acd000)
	libgomp.so.1 => /usr/lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f9272a9a000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f9272a80000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f92728bf000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f927306b000)
```

**Comment by PeerHerholz** [comment]
hi @kaczmarj, 

thank you very much for following up on this and the input.

I started playing around with it today, trying to get it to run in a minimal `docker image`.

**Comment by PeerHerholz** [comment]
Hi gang, 

I was wondering if we should implement the functionality as discussed or actually support volume formats?
I'm asking because I think a `surface` to `volume` transformation would be rather limited to visualization purposes given
that the random parcellation was done on the surface and is then transformed (which no matter how you run it will be suboptimal). @miykael how rough would the support of voxels be?

**Comment by miykael** [comment]
Hey @PeerHerholz, thank you for picking this up again.

**Implementing functionality as discussed** seems tricky, for the reasons mentioned above. The ideal would be to keep the toolbox in pure Python code. But as you mention, `nibabel` or `nilearn` don't seem to be able to do the `surf2vol` projection. However, could we potentially mirror/inverse the transformation used in `nilearn`'s [vol_to_surf](https://nilearn.github.io/modules/generated/nilearn.surface.vol_to_surf.html#nilearn.surface.vol_to_surf) function? The code would start [here](https://github.com/nilearn/nilearn/blob/master/nilearn/surface/surface.py#L388). Personally, I believe that would lead to too much work.

With the **support of voxels**, do you mean for example to have a 3D binary mask and to subdivide it into *N* equal segments? Looking at our source code and if I remember correctly, I believe that the crucial parts of the toolbox don't require the input to be on the surface. The locations of the vertices that are used to compute adjacency and distance could also be based on the location of the voxels. I believe only, the [colormaps.py](https://github.com/miykael/parcellation_fragmenter/blob/master/fragmenter/colormaps.py) and the [Rotator.py](https://github.com/miykael/parcellation_fragmenter/blob/master/fragmenter/Rotator.py) script would need the surface to be on the sphere.  
If I remember correctly, a volume also has less voxels than a surface has vertices, so computational demand should be reasonable. However, segmentation in the volume might lead to some issues if the input volume is only restricted on a certain tissue type (e.g. gray matter). Or in other words, what if the center of a segment is in a ventricle? Such a segment might include all relevant voxels around the ventricle, but nothing is really connected.

Also, in what way would such an approach improve on what Cameron Craddock already did in 2011 ([A Whole Brain fMRI Atlas Generated via Spatially Constrained Spectral Clustering](http://birc.jaredjustus.com/assets/publications/Craddock,%20James%202011%20A%20whole%20brain%20fMRI%20atlas.pdf)), or Jean-Baptiste Poline in 2014 ([Which fMRI clustering gives good brain parcellations
](https://core.ac.uk/download/pdf/82866511.pdf))?
